pages/en/lb4/Testing-Your-application.md

```diff
@@ -8,676 +8,556 @@ permalink: /doc/en/lb4/Testing-your-application.html
+npm install --save-dev mocha @loopback/testlab
 ```
 
 Your `package.json` should look something like:
 
 ```js
-  ...
+{
+  // ...
   "devDependencies": {
-    "@loopback/testlab": "<current-version>",
-    "mocha": "<current-version>"
+    "@loopback/testlab": "^<current-version>",
+    "mocha": "^<current-version>"
   },
   "scripts": {
     "test": "mocha"
   }
-  ...
+  // ...
+}
 ```
 
-## Acceptance testing
+## Data handling
 
-Create an application that returns "hello world" for GET / requests.
+Tests accessing a real database often require pre-existing data (e.g. a method listing all products need some products to exist in the database) and/or need to create new model instances with valid data (e.g a method creating a new product need to know how to fill in all required properties). There are various approaches for solving these two problems, many of them unfortunately make the test suite difficult to understand, difficult to maintain and prone to test failures unrelated to the changes made.
 
-Start by writing an acceptance test that ensures the application you're creating returns "hello world". Create the `test` directory and a file inside that looks like:
+Based on our experience, we recommend the following approach.
 
-{% include code-caption.html content= "test/acceptance.js" %}
-```js
+### Clean the database before each test
 
-const MyApp = require('../application');
-const createHandler = require('@loopback/testlab').createClientForRestServer;
-const expect = require('@loopback/testlab').expect;
-const RestServer = require('@loopback/rest').RestServer;
-
-describe('application', () => {
-  let app;
-  let client;
-  before(createClient);
-  after(stopApp);
-
-  it('returns "hello world" for GET requests to /', async () => {
-    const res = await client.get('/');
-    expect(res.statusCode).to.eql(200);
-    expect(res.text).to.eql('hello world');
-  });
+Always start with a clean database before each test. This may seem counter-intuitive: why not reset the database after the test has finished? When a test fails and the database is cleaned after the test has finished, then it's difficult to observe what was stored in the database, why the test failed. When the database is cleaned in the beginning, then any failing test will leave the database exactly in the state that caused the test to fail.
 
-  // Create our supertest instance and start up the application (and server!)
-  async function createClient() {
-    app = new MyApp();
-    const server = await app.getServer(RestServer);
-    await app.start();
-    client = await createHandler(server);
-  }
+This is typically achieved by setting up a `beforeEach` hook calling a helper method to clean the database.
 
-  // Shut down the application and its server instance.
-  async function stopApp() {
-    await app.stop();
-  }
+
+```ts
+// test/helpers/database.helpers.ts
+
+export async function givenEmptyDatabase() {
+  await new ProductRepository().deleteAll();
```

Comment:
Would it be better to recommend using docker to setup a clean database?

- from my experience api like `deleteAll()` some times also interferes testing result/pollute db if it has bug and could not clean up the old data entirely. 

- you can also leave the container there instead of deleting it after test :)

I feel we benefit a lot after lb3 connectors all use container for testing.

Reply:
We are running `givenEmptyDatabase` before each test. How long does it take to restart a docker container with a database? I am concerned that cleaning the database via docker restart would make the test suite extremely slow.

> from my experience api like deleteAll() some times also interferes testing result/pollute db if it has bug and could not clean up the old data entirely.

I see your point. This can happen for example when there are foreign keys setup between tables and let' say a `Category` record cannot be deleted until all `Product` records belonging to that category are deleted too. IMO, this is a missing feature of LoopBack - we should support cascading deletes. Fortunately, this particular example has a workaround - just re-order `deleteAll` statements so that products are deleted before categories.

I think we shouldn't be recommending our users suboptimal solutions just to avoid shortcomings in LoopBack and/or the way how the users designed their app & data schema. What I am proposing instead: let's use this as an opportunity to identify missing gaps and either explain users how to work around them, or work together to fix these issues.

@jannyHou What do you think, should I add a short paragraph explaining how to clear database when there are relationships between tables enforced by the DB engine, like I shown in my first paragraph?

Reply:
@bajtos yeah the time is a problem for starting a container...

I think we can add more after figure out a plan for persistence story, now your content is good enough, going into too much details might be off topic.